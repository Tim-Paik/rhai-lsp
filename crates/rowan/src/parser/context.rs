//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
