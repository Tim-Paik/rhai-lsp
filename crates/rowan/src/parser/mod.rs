//! This module contains all parsing-related tools including a
//! recursive-descent parser.

use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

/// A flexible parser.
/// 
/// Parsing happens via given parser functions (also found in [`parsers`]).
/// 
/// # Example Uses
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// // Parse some Rhai code (file).
/// let parser = Parser::new(r#"print("hello rhai!")"#);
/// let parse = parser.parse();
/// 
/// assert!(parse.errors.is_empty());
/// ```
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// use rhai_rowan::parser::parsers::parse_expr;
///
/// // Parse an expression explicitly.
/// let mut parser = Parser::new(r#"print("hello rhai!")"#);
/// parser.execute(parse_expr);
/// let parse = parser.finish();
/// 
/// assert!(parse.errors.is_empty());
/// ```
pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    /// Create a new parser with the given source.
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    /// Finish parsing.
    /// 
    /// # Panics
    /// 
    /// If no parser function was called,
    /// or a parser was left in an invalid state (bug).
    #[must_use]
    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    /// Execute a parser function for the given parser.
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context);
    }
}

/// The result of parsing, containing all errors
/// and the final green tree.
#[derive(Debug, Clone)]
pub struct Parse {
    /// Parse errors.
    pub errors: Vec<ParseError>,
    /// Parsed green tree.
    pub green: GreenNode,
}

impl Parse {
    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn clone_syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }

    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

/// A parse (syntax) error.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    /// The span of the error in the parsed source.
    pub range: rowan::TextRange,
    /// Error kind.
    pub kind: ParseErrorKind,
}

impl ParseError {
    fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

/// All the non-fatal parsing errors that can occur.
#[derive(Debug, PartialEq, Eq, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .collect::<Vec<String>>().join(",")
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
