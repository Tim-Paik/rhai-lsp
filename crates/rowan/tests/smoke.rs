use rhai_rowan::parser::{Operator, Parser};
use test_case::test_case;

#[test_case("simple", include_str!("../../../testdata/valid/simple.rhai"))]
#[test_case("array", include_str!("../../../testdata/valid/array.rhai"))]
#[test_case("assignment", include_str!("../../../testdata/valid/assignment.rhai"))]
#[test_case("comments", include_str!("../../../testdata/valid/comments.rhai"))]
#[test_case("fibonacci", include_str!("../../../testdata/valid/fibonacci.rhai"))]
#[test_case("for1", include_str!("../../../testdata/valid/for1.rhai"))]
#[test_case("for2", include_str!("../../../testdata/valid/for2.rhai"))]
#[test_case("function_decl1", include_str!("../../../testdata/valid/function_decl1.rhai"))]
#[test_case("function_decl2", include_str!("../../../testdata/valid/function_decl2.rhai"))]
#[test_case("function_decl3", include_str!("../../../testdata/valid/function_decl3.rhai"))]
#[test_case("function_decl4", include_str!("../../../testdata/valid/function_decl4.rhai"))]
#[test_case("if1", include_str!("../../../testdata/valid/if1.rhai"))]
#[test_case("if2", include_str!("../../../testdata/valid/if2.rhai"))]
#[test_case("loop", include_str!("../../../testdata/valid/loop.rhai"))]
#[test_case("mat_mul", include_str!("../../../testdata/valid/mat_mul.rhai"))]
#[test_case("module", include_str!("../../../testdata/valid/module.rhai"))]
#[test_case("oop", include_str!("../../../testdata/valid/oop.rhai"))]
#[test_case("op1", include_str!("../../../testdata/valid/op1.rhai"))]
#[test_case("op2", include_str!("../../../testdata/valid/op2.rhai"))]
#[test_case("op3", include_str!("../../../testdata/valid/op3.rhai"))]
#[test_case("primes", include_str!("../../../testdata/valid/primes.rhai"))]
#[test_case("speed_test", include_str!("../../../testdata/valid/speed_test.rhai"))]
#[test_case("string", include_str!("../../../testdata/valid/string.rhai"))]
#[test_case("strings_map", include_str!("../../../testdata/valid/strings_map.rhai"))]
#[test_case("switch", include_str!("../../../testdata/valid/switch.rhai"))]
#[test_case("while", include_str!("../../../testdata/valid/while.rhai"))]
#[test_case("char", include_str!("../../../testdata/valid/char.rhai"))]
#[test_case("throw_try_catch", include_str!("../../../testdata/valid/throw_try_catch.rhai"))]
#[test_case("optional_ops", include_str!("../../../testdata/valid/optional_ops.rhai"))]
fn parse_valid(name: &str, src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty(), "{:#?}", parse.errors);

    let mut engine = rhai::Engine::new();
    engine.set_max_expr_depths(0, 0);

    if src.starts_with("#!") {
        engine
            .compile(
                src.lines()
                    .skip(1)
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join("\n"),
            )
            .unwrap();
    } else {
        engine.compile(src).unwrap();
    }

    insta::with_settings!(
        { snapshot_suffix => name },
        {
            insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
        }
    );
}

#[test]
fn parse_custom_operator() {
    let src = include_str!("../../../testdata/valid/operators.rhai");

    let parse = Parser::new(src)
        .with_operator("over", Operator::default())
        .parse_script();
    assert!(parse.errors.is_empty(), "{:#?}", parse.errors);

    insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));

    let parse = Parser::new(src).parse_script();
    assert!(!parse.errors.is_empty());
}
