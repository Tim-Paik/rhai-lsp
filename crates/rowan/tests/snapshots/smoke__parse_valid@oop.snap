---
source: crates/rowan/tests/smoke.rs
assertion_line: 37
expression: "format!(\"{:#?}\", parse.into_syntax())"
---
RHAI@0..1409
  COMMENT_LINE@0..85 "// This script simula ..."
  WHITESPACE@85..87 "\n\n"
  COMMENT_LINE@87..130 "// External variable  ..."
  WHITESPACE@130..131 "\n"
  STMT@131..151
    ITEM@131..150
      EXPR@131..150
        EXPR_LET@131..150
          KW_LET@131..134 "let"
          WHITESPACE@134..135 " "
          IDENT@135..145 "last_value"
          WHITESPACE@145..146 " "
          OP_ASSIGN@146..147 "="
          EXPR@147..150
            WHITESPACE@147..148 " "
            EXPR_PAREN@148..150
              PUNCT_PAREN_START@148..149 "("
              PUNCT_PAREN_END@149..150 ")"
    PUNCT_SEMI@150..151 ";"
  WHITESPACE@151..153 "\n\n"
  COMMENT_LINE@153..169 "// Define object"
  WHITESPACE@169..170 "\n"
  STMT@170..565
    ITEM@170..564
      EXPR@170..564
        EXPR_LET@170..564
          KW_LET@170..173 "let"
          WHITESPACE@173..174 " "
          IDENT@174..178 "obj1"
          WHITESPACE@178..179 " "
          OP_ASSIGN@179..180 "="
          EXPR@180..564
            WHITESPACE@180..181 " "
            EXPR_OBJECT@181..564
              PUNCT_MAP_START@181..183 "#{"
              WHITESPACE@183..188 "\n    "
              OBJECT_FIELD@188..197
                IDENT@188..193 "_data"
                PUNCT_COLON@193..194 ":"
                EXPR@194..197
                  WHITESPACE@194..195 " "
                  EXPR_LIT@195..197
                    LIT@195..197
                      LIT_INT@195..197 "42"
              PUNCT_COMMA@197..198 ","
              WHITESPACE@198..232 "                      ..."
              COMMENT_LINE@232..245 "// data field"
              WHITESPACE@245..250 "\n    "
              OBJECT_FIELD@250..273
                IDENT@250..258 "get_data"
                PUNCT_COLON@258..259 ":"
                EXPR@259..273
                  WHITESPACE@259..260 " "
                  EXPR_CLOSURE@260..273
                    PARAM_LIST@260..262
                      OP_BIT_OR@260..261 "|"
                      OP_BIT_OR@261..262 "|"
                    EXPR@262..273
                      EXPR_BINARY@262..273
                        EXPR@262..267
                          WHITESPACE@262..263 " "
                          EXPR_IDENT@263..267
                            IDENT@263..267 "this"
                        PUNCT_DOT@267..268 "."
                        EXPR@268..273
                          EXPR_IDENT@268..273
                            IDENT@268..273 "_data"
              PUNCT_COMMA@273..274 ","
              WHITESPACE@274..294 "                    "
              COMMENT_LINE@294..312 "// property getter"
              WHITESPACE@312..317 "\n    "
              OBJECT_FIELD@317..355
                IDENT@317..323 "action"
                PUNCT_COLON@323..324 ":"
                EXPR@324..355
                  WHITESPACE@324..325 " "
                  EXPR_CLOSURE@325..355
                    PARAM_LIST@325..327
                      OP_BIT_OR@325..326 "|"
                      OP_BIT_OR@326..327 "|"
                    EXPR@327..355
                      EXPR_CALL@327..355
                        EXPR@327..333
                          WHITESPACE@327..328 " "
                          EXPR_IDENT@328..333
                            IDENT@328..333 "print"
                        ARG_LIST@333..355
                          PUNCT_PAREN_START@333..334 "("
                          EXPR@334..354
                            EXPR_LIT@334..354
                              LIT@334..354
                                LIT_STR@334..354 "`Data=${this._data}`"
                          PUNCT_PAREN_END@354..355 ")"
              PUNCT_COMMA@355..356 ","
              WHITESPACE@356..361 "     "
              COMMENT_LINE@361..370 "// method"
              WHITESPACE@370..375 "\n    "
              OBJECT_FIELD@375..563
                IDENT@375..381 "update"
                PUNCT_COLON@381..382 ":"
                EXPR@382..563
                  WHITESPACE@382..383 " "
                  EXPR_CLOSURE@383..563
                    PARAM_LIST@383..386
                      OP_BIT_OR@383..384 "|"
                      PARAM@384..385
                        IDENT@384..385 "x"
                      OP_BIT_OR@385..386 "|"
                    EXPR@386..563
                      WHITESPACE@386..387 " "
                      EXPR_BLOCK@387..562
                        PUNCT_BRACE_START@387..388 "{"
                        WHITESPACE@388..419 "                      ..."
                        COMMENT_LINE@419..437 "// property setter"
                        WHITESPACE@437..446 "\n        "
                        STMT@446..461
                          ITEM@446..460
                            EXPR@446..460
                              EXPR_BINARY@446..460
                                EXPR@446..457
                                  EXPR_BINARY@446..457
                                    EXPR@446..450
                                      EXPR_IDENT@446..450
                                        IDENT@446..450 "this"
                                    PUNCT_DOT@450..451 "."
                                    EXPR@451..457
                                      EXPR_IDENT@451..457
                                        IDENT@451..456 "_data"
                                        WHITESPACE@456..457 " "
                                OP_ASSIGN@457..458 "="
                                EXPR@458..460
                                  WHITESPACE@458..459 " "
                                  EXPR_IDENT@459..460
                                    IDENT@459..460 "x"
                          PUNCT_SEMI@460..461 ";"
                        WHITESPACE@461..470 "\n        "
                        STMT@470..494
                          ITEM@470..493
                            EXPR@470..493
                              EXPR_BINARY@470..493
                                EXPR@470..481
                                  EXPR_IDENT@470..481
                                    IDENT@470..480 "last_value"
                                    WHITESPACE@480..481 " "
                                OP_ASSIGN@481..482 "="
                                EXPR@482..493
                                  EXPR_BINARY@482..493
                                    EXPR@482..487
                                      WHITESPACE@482..483 " "
                                      EXPR_IDENT@483..487
                                        IDENT@483..487 "this"
                                    PUNCT_DOT@487..488 "."
                                    EXPR@488..493
                                      EXPR_IDENT@488..493
                                        IDENT@488..493 "_data"
                          PUNCT_SEMI@493..494 ";"
                        WHITESPACE@494..510 "                "
                        COMMENT_LINE@510..533 "// capture 'last_value'"
                        WHITESPACE@533..542 "\n        "
                        STMT@542..556
                          ITEM@542..555
                            EXPR@542..555
                              EXPR_CALL@542..555
                                EXPR@542..553
                                  EXPR_BINARY@542..553
                                    EXPR@542..546
                                      EXPR_IDENT@542..546
                                        IDENT@542..546 "this"
                                    PUNCT_DOT@546..547 "."
                                    EXPR@547..553
                                      EXPR_IDENT@547..553
                                        IDENT@547..553 "action"
                                ARG_LIST@553..555
                                  PUNCT_PAREN_START@553..554 "("
                                  PUNCT_PAREN_END@554..555 ")"
                          PUNCT_SEMI@555..556 ";"
                        WHITESPACE@556..561 "\n    "
                        PUNCT_BRACE_END@561..562 "}"
                      WHITESPACE@562..563 "\n"
              PUNCT_BRACE_END@563..564 "}"
    PUNCT_SEMI@564..565 ";"
  WHITESPACE@565..567 "\n\n"
  STMT@567..797
    ITEM@567..797
      EXPR@567..797
        EXPR_IF@567..744
          KW_IF@567..569 "if"
          EXPR@569..590
            EXPR_BINARY@569..590
              EXPR@569..586
                EXPR_CALL@569..585
                  EXPR@569..583
                    EXPR_BINARY@569..583
                      EXPR@569..574
                        WHITESPACE@569..570 " "
                        EXPR_IDENT@570..574
                          IDENT@570..574 "obj1"
                      PUNCT_DOT@574..575 "."
                      EXPR@575..583
                        EXPR_IDENT@575..583
                          IDENT@575..583 "get_data"
                  ARG_LIST@583..585
                    PUNCT_PAREN_START@583..584 "("
                    PUNCT_PAREN_END@584..585 ")"
                WHITESPACE@585..586 " "
              OP_GT@586..587 ">"
              EXPR@587..590
                WHITESPACE@587..588 " "
                EXPR_LIT@588..589
                  LIT@588..589
                    LIT_INT@588..589 "0"
                WHITESPACE@589..590 " "
          EXPR_BLOCK@590..698
            PUNCT_BRACE_START@590..591 "{"
            WHITESPACE@591..615 "                        "
            COMMENT_LINE@615..633 "// property access"
            WHITESPACE@633..638 "\n    "
            STMT@638..655
              ITEM@638..654
                EXPR@638..654
                  EXPR_CALL@638..654
                    EXPR@638..649
                      EXPR_BINARY@638..649
                        EXPR@638..642
                          EXPR_IDENT@638..642
                            IDENT@638..642 "obj1"
                        PUNCT_DOT@642..643 "."
                        EXPR@643..649
                          EXPR_IDENT@643..649
                            IDENT@643..649 "update"
                    ARG_LIST@649..654
                      PUNCT_PAREN_START@649..650 "("
                      EXPR@650..653
                        EXPR_LIT@650..653
                          LIT@650..653
                            LIT_INT@650..653 "123"
                      PUNCT_PAREN_END@653..654 ")"
              PUNCT_SEMI@654..655 ";"
            WHITESPACE@655..682 "                      ..."
            COMMENT_LINE@682..696 "// call method"
            WHITESPACE@696..697 "\n"
            PUNCT_BRACE_END@697..698 "}"
          WHITESPACE@698..699 " "
          KW_ELSE@699..703 "else"
          WHITESPACE@703..704 " "
          EXPR_BLOCK@704..744
            PUNCT_BRACE_START@704..705 "{"
            WHITESPACE@705..710 "\n    "
            STMT@710..742
              ITEM@710..741
                EXPR@710..741
                  EXPR_CALL@710..741
                    EXPR@710..715
                      EXPR_IDENT@710..715
                        IDENT@710..715 "print"
                    ARG_LIST@715..741
                      PUNCT_PAREN_START@715..716 "("
                      EXPR@716..740
                        EXPR_LIT@716..740
                          LIT@716..740
                            LIT_STR@716..740 "\"we have a problem here\""
                      PUNCT_PAREN_END@740..741 ")"
              PUNCT_SEMI@741..742 ";"
            WHITESPACE@742..743 "\n"
            PUNCT_BRACE_END@743..744 "}"
        WHITESPACE@744..746 "\n\n"
        COMMENT_LINE@746..796 "// Define another obj ..."
        WHITESPACE@796..797 "\n"
  STMT@797..1102
    ITEM@797..1101
      EXPR@797..1101
        EXPR_LET@797..1101
          KW_LET@797..800 "let"
          WHITESPACE@800..801 " "
          IDENT@801..805 "obj2"
          WHITESPACE@805..806 " "
          OP_ASSIGN@806..807 "="
          EXPR@807..1101
            WHITESPACE@807..808 " "
            EXPR_OBJECT@808..1101
              PUNCT_MAP_START@808..810 "#{"
              WHITESPACE@810..815 "\n    "
              OBJECT_FIELD@815..823
                IDENT@815..820 "_data"
                PUNCT_COLON@820..821 ":"
                EXPR@821..823
                  WHITESPACE@821..822 " "
                  EXPR_LIT@822..823
                    LIT@822..823
                      LIT_INT@822..823 "0"
              PUNCT_COMMA@823..824 ","
              WHITESPACE@824..859 "                      ..."
              COMMENT_LINE@859..884 "// data field - new v ..."
              WHITESPACE@884..889 "\n    "
              OBJECT_FIELD@889..1100
                IDENT@889..895 "update"
                PUNCT_COLON@895..896 ":"
                EXPR@896..1100
                  WHITESPACE@896..897 " "
                  EXPR_CLOSURE@897..1100
                    PARAM_LIST@897..900
                      OP_BIT_OR@897..898 "|"
                      PARAM@898..899
                        IDENT@898..899 "x"
                      OP_BIT_OR@899..900 "|"
                    EXPR@900..1100
                      WHITESPACE@900..901 " "
                      EXPR_BLOCK@901..1099
                        PUNCT_BRACE_START@901..902 "{"
                        WHITESPACE@902..933 "                      ..."
                        COMMENT_LINE@933..970 "// property setter -  ..."
                        WHITESPACE@970..979 "\n        "
                        STMT@979..998
                          ITEM@979..997
                            EXPR@979..997
                              EXPR_BINARY@979..997
                                EXPR@979..990
                                  EXPR_BINARY@979..990
                                    EXPR@979..983
                                      EXPR_IDENT@979..983
                                        IDENT@979..983 "this"
                                    PUNCT_DOT@983..984 "."
                                    EXPR@984..990
                                      EXPR_IDENT@984..990
                                        IDENT@984..989 "_data"
                                        WHITESPACE@989..990 " "
                                OP_ASSIGN@990..991 "="
                                EXPR@991..997
                                  EXPR_BINARY@991..997
                                    EXPR@991..994
                                      WHITESPACE@991..992 " "
                                      EXPR_IDENT@992..994
                                        IDENT@992..993 "x"
                                        WHITESPACE@993..994 " "
                                    OP_MUL@994..995 "*"
                                    EXPR@995..997
                                      WHITESPACE@995..996 " "
                                      EXPR_LIT@996..997
                                        LIT@996..997
                                          LIT_INT@996..997 "2"
                          PUNCT_SEMI@997..998 ";"
                        WHITESPACE@998..1007 "\n        "
                        STMT@1007..1031
                          ITEM@1007..1030
                            EXPR@1007..1030
                              EXPR_BINARY@1007..1030
                                EXPR@1007..1018
                                  EXPR_IDENT@1007..1018
                                    IDENT@1007..1017 "last_value"
                                    WHITESPACE@1017..1018 " "
                                OP_ASSIGN@1018..1019 "="
                                EXPR@1019..1030
                                  EXPR_BINARY@1019..1030
                                    EXPR@1019..1024
                                      WHITESPACE@1019..1020 " "
                                      EXPR_IDENT@1020..1024
                                        IDENT@1020..1024 "this"
                                    PUNCT_DOT@1024..1025 "."
                                    EXPR@1025..1030
                                      EXPR_IDENT@1025..1030
                                        IDENT@1025..1030 "_data"
                          PUNCT_SEMI@1030..1031 ";"
                        WHITESPACE@1031..1047 "                "
                        COMMENT_LINE@1047..1070 "// capture 'last_value'"
                        WHITESPACE@1070..1079 "\n        "
                        STMT@1079..1093
                          ITEM@1079..1092
                            EXPR@1079..1092
                              EXPR_CALL@1079..1092
                                EXPR@1079..1090
                                  EXPR_BINARY@1079..1090
                                    EXPR@1079..1083
                                      EXPR_IDENT@1079..1083
                                        IDENT@1079..1083 "this"
                                    PUNCT_DOT@1083..1084 "."
                                    EXPR@1084..1090
                                      EXPR_IDENT@1084..1090
                                        IDENT@1084..1090 "action"
                                ARG_LIST@1090..1092
                                  PUNCT_PAREN_START@1090..1091 "("
                                  PUNCT_PAREN_END@1091..1092 ")"
                          PUNCT_SEMI@1092..1093 ";"
                        WHITESPACE@1093..1098 "\n    "
                        PUNCT_BRACE_END@1098..1099 "}"
                      WHITESPACE@1099..1100 "\n"
              PUNCT_BRACE_END@1100..1101 "}"
    PUNCT_SEMI@1101..1102 ";"
  WHITESPACE@1102..1103 "\n"
  STMT@1103..1124
    ITEM@1103..1123
      EXPR@1103..1123
        EXPR_CALL@1103..1123
          EXPR@1103..1117
            EXPR_BINARY@1103..1117
              EXPR@1103..1107
                EXPR_IDENT@1103..1107
                  IDENT@1103..1107 "obj2"
              PUNCT_DOT@1107..1108 "."
              EXPR@1108..1117
                EXPR_IDENT@1108..1117
                  IDENT@1108..1117 "fill_with"
          ARG_LIST@1117..1123
            PUNCT_PAREN_START@1117..1118 "("
            EXPR@1118..1122
              EXPR_IDENT@1118..1122
                IDENT@1118..1122 "obj1"
            PUNCT_PAREN_END@1122..1123 ")"
    PUNCT_SEMI@1123..1124 ";"
  WHITESPACE@1124..1151 "                      ..."
  COMMENT_LINE@1151..1184 "// add all other fiel ..."
  WHITESPACE@1184..1186 "\n\n"
  STMT@1186..1371
    ITEM@1186..1371
      EXPR@1186..1371
        EXPR_IF@1186..1369
          KW_IF@1186..1188 "if"
          EXPR@1188..1209
            EXPR_BINARY@1188..1209
              EXPR@1188..1205
                EXPR_CALL@1188..1204
                  EXPR@1188..1202
                    EXPR_BINARY@1188..1202
                      EXPR@1188..1193
                        WHITESPACE@1188..1189 " "
                        EXPR_IDENT@1189..1193
                          IDENT@1189..1193 "obj2"
                      PUNCT_DOT@1193..1194 "."
                      EXPR@1194..1202
                        EXPR_IDENT@1194..1202
                          IDENT@1194..1202 "get_data"
                  ARG_LIST@1202..1204
                    PUNCT_PAREN_START@1202..1203 "("
                    PUNCT_PAREN_END@1203..1204 ")"
                WHITESPACE@1204..1205 " "
              OP_GT@1205..1206 ">"
              EXPR@1206..1209
                WHITESPACE@1206..1207 " "
                EXPR_LIT@1207..1208
                  LIT@1207..1208
                    LIT_INT@1207..1208 "0"
                WHITESPACE@1208..1209 " "
          EXPR_BLOCK@1209..1297
            PUNCT_BRACE_START@1209..1210 "{"
            WHITESPACE@1210..1234 "                        "
            COMMENT_LINE@1234..1252 "// property access"
            WHITESPACE@1252..1257 "\n    "
            STMT@1257..1295
              ITEM@1257..1294
                EXPR@1257..1294
                  EXPR_CALL@1257..1294
                    EXPR@1257..1262
                      EXPR_IDENT@1257..1262
                        IDENT@1257..1262 "print"
                    ARG_LIST@1262..1294
                      PUNCT_PAREN_START@1262..1263 "("
                      EXPR@1263..1293
                        EXPR_LIT@1263..1293
                          LIT@1263..1293
                            LIT_STR@1263..1293 "\"we have another prob ..."
                      PUNCT_PAREN_END@1293..1294 ")"
              PUNCT_SEMI@1294..1295 ";"
            WHITESPACE@1295..1296 "\n"
            PUNCT_BRACE_END@1296..1297 "}"
          WHITESPACE@1297..1298 " "
          KW_ELSE@1298..1302 "else"
          WHITESPACE@1302..1303 " "
          EXPR_BLOCK@1303..1369
            PUNCT_BRACE_START@1303..1304 "{"
            WHITESPACE@1304..1309 "\n    "
            STMT@1309..1325
              ITEM@1309..1324
                EXPR@1309..1324
                  EXPR_CALL@1309..1324
                    EXPR@1309..1320
                      EXPR_BINARY@1309..1320
                        EXPR@1309..1313
                          EXPR_IDENT@1309..1313
                            IDENT@1309..1313 "obj2"
                        PUNCT_DOT@1313..1314 "."
                        EXPR@1314..1320
                          EXPR_IDENT@1314..1320
                            IDENT@1314..1320 "update"
                    ARG_LIST@1320..1324
                      PUNCT_PAREN_START@1320..1321 "("
                      EXPR@1321..1323
                        EXPR_LIT@1321..1323
                          LIT@1321..1323
                            LIT_INT@1321..1323 "42"
                      PUNCT_PAREN_END@1323..1324 ")"
              PUNCT_SEMI@1324..1325 ";"
            WHITESPACE@1325..1353 "                      ..."
            COMMENT_LINE@1353..1367 "// call method"
            WHITESPACE@1367..1368 "\n"
            PUNCT_BRACE_END@1368..1369 "}"
        WHITESPACE@1369..1371 "\n\n"
  STMT@1371..1408
    ITEM@1371..1407
      EXPR@1371..1407
        EXPR_CALL@1371..1407
          EXPR@1371..1376
            EXPR_IDENT@1371..1376
              IDENT@1371..1376 "print"
          ARG_LIST@1376..1407
            PUNCT_PAREN_START@1376..1377 "("
            EXPR@1377..1406
              EXPR_LIT@1377..1406
                LIT@1377..1406
                  LIT_STR@1377..1406 "`Should be 84: ${last ..."
            PUNCT_PAREN_END@1406..1407 ")"
    PUNCT_SEMI@1407..1408 ";"
  WHITESPACE@1408..1409 "\n"

